File: php/src/Runner.php
<?php

namespace Call\Coding\Clis;

class Runner
{
    private $executor;

    public function __construct(?callable $executor = null)
    {
        $this->executor = $executor ?? self::class . '::defaultRun';
    }

    public function run(CommandSpec $spec): CompletedRun
    {
        return ($this->executor)($spec);
    }

    public function stream(CommandSpec $spec, callable $onEvent): CompletedRun
    {
        $result = $this->run($spec);
        if ($result->stdout !== '') {
            $onEvent('stdout', $result->stdout);
        }
        if ($result->stderr !== '') {
            $onEvent('stderr', $result->stderr);
        }
        return $result;
    }

    public static function defaultRun(CommandSpec $spec): CompletedRun
    {
        $argv = $spec->argv;
        $argv0 = $argv[0];

        $env = array_merge(getenv(), $spec->env);

        $descriptors = [
            0 => ['pipe', 'r'],
            1 => ['pipe', 'w'],
            2 => ['pipe', 'w'],
        ];

        $origCwd = null;
        if ($spec->cwd !== null) {
            $origCwd = getcwd();
        }

        try {
            if ($spec->cwd !== null) {
                chdir($spec->cwd);
            }

            $proc = @proc_open($argv, $descriptors, $pipes, null, $env);

            if ($proc === false) {
                $error = error_get_last()['message'] ?? 'unknown error';
                self::restoreCwd($origCwd);
                return new CompletedRun($argv, 1, '', "failed to start $argv0: $error\n");
            }

            if ($spec->stdin_text !== null && strlen($spec->stdin_text) > 0) {
                fwrite($pipes[0], $spec->stdin_text);
            }
            fclose($pipes[0]);

            $stdout = stream_get_contents($pipes[1]);
            $stderr = stream_get_contents($pipes[2]);
            fclose($pipes[1]);
            fclose($pipes[2]);

            $exitCode = proc_close($proc);

            self::restoreCwd($origCwd);

            return new CompletedRun($argv, $exitCode, $stdout ?? '', $stderr ?? '');
        } catch (\Throwable $e) {
            self::restoreCwd($origCwd);
            $err = rtrim($e->getMessage());
            return new CompletedRun($argv, 1, '', "failed to start $argv0: $err\n");
        }
    }

    private static function restoreCwd(?string $origCwd): void
    {
        if ($origCwd !== null) {
            chdir($origCwd);
        }
    }
}

File: zig/src/runner.zig
const std = @import("std");

pub const CommandSpec = struct {
    argv: []const []const u8,
    stdin_text: ?[]const u8 = null,
    cwd: ?[]const u8 = null,
};

pub const CompletedRun = struct {
    allocator: std.mem.Allocator,
    exit_code: u8,
    stdout_data: []const u8,
    stderr_data: []const u8,

    pub fn deinit(self: *const CompletedRun) void {
        self.allocator.free(self.stdout_data);
        self.allocator.free(self.stderr_data);
    }
};

pub const Runner = struct {
    const max_output: usize = 10 * 1024 * 1024;

    fn failResult(allocator: std.mem.Allocator, argv0: []const u8, err: anyerror) !CompletedRun {
        return CompletedRun{
            .allocator = allocator,
            .exit_code = 1,
            .stdout_data = try allocator.dupe(u8, ""),
            .stderr_data = std.fmt.allocPrint(allocator, "failed to start {s}: {s}\n", .{ argv0, @errorName(err) }) catch unreachable,
        };
    }

    pub fn run(allocator: std.mem.Allocator, spec: CommandSpec) !CompletedRun {
        var child = std.process.Child.init(spec.argv, allocator);
        child.stdin_behavior = .Pipe;
        child.stdout_behavior = .Pipe;
        child.stderr_behavior = .Pipe;
        if (spec.cwd) |cwd_path| {
            child.cwd = cwd_path;
        }

        child.spawn() catch |err| {
            return failResult(allocator, spec.argv[0], err);
        };

        if (spec.stdin_text) |text| {
            child.stdin.?.writeAll(text) catch {};
        }
        child.stdin.?.close();
        child.stdin = null;

        var stdout_buf: std.ArrayList(u8) = .empty;
        defer stdout_buf.deinit(allocator);
        var stderr_buf: std.ArrayList(u8) = .empty;
        defer stderr_buf.deinit(allocator);

        child.collectOutput(allocator, &stdout_buf, &stderr_buf, max_output) catch {};

        const term = child.wait() catch |err| {
            return failResult(allocator, spec.argv[0], err);
        };

        const exit_code: u8 = switch (term) {
            .Exited => |code| code,
            else => 1,
        };

        return CompletedRun{
            .allocator = allocator,
            .exit_code = exit_code,
            .stdout_data = try stdout_buf.toOwnedSlice(allocator),
            .stderr_data = try stderr_buf.toOwnedSlice(allocator),
        };
    }

    pub fn stream(allocator: std.mem.Allocator, spec: CommandSpec, on_event: *const fn ([]const u8, []const u8) void) !CompletedRun {
        const result = try run(allocator, spec);
        if (result.stdout_data.len > 0) on_event("stdout", result.stdout_data);
        if (result.stderr_data.len > 0) on_event("stderr", result.stderr_data);
        return result;
    }
};
